benchmark/spmv/batch_spmv.cpp

```diff
@@ -0,0 +1,270 @@
+        gko::_tuning_flag = false;
+#endif  // GINKGO_BENCHMARK_ENABLE_TUNING
+
+        // timed run
+        auto timer = get_timer(exec, FLAGS_gpu_timer);
+        for (unsigned int i = 0; i < FLAGS_repetitions; i++) {
+            auto x_clone = clone(x);
+            exec->synchronize();
+            timer->tic();
+            system_matrix->apply(lend(b), lend(x_clone));
+            timer->toc();
+        }
+        add_or_set_member(spmv_case[format_name], "time",
+                          timer->compute_average_time(), allocator);
+
+        // compute and write benchmark data
+        add_or_set_member(spmv_case[format_name], "completed", true, allocator);
+    } catch (const std::exception &e) {
+        add_or_set_member(test_case["spmv"][format_name], "completed", false,
+                          allocator);
+        std::cerr << "Error when processing test case " << test_case << "\n"
+                  << "what(): " << e.what() << std::endl;
+    }
+}
+
+
+int main(int argc, char *argv[])
+{
+    std::string header =
+        "A benchmark for measuring performance of Ginkgo's spmv.\n";
+    std::string format = std::string() + "  [\n" +
+                         "    { \"filename\": \"my_file.mtx\"},\n" +
+                         "    { \"filename\": \"my_file2.mtx\"}\n" + "  ]\n\n";
+    initialize_argument_parsing(&argc, &argv, header, format);
+
+    std::string extra_information = "The formats are " + FLAGS_formats +
+                                    "\nThe number of right hand sides is " +
+                                    std::to_string(FLAGS_nrhs) + "\n";
+    print_general_information(extra_information);
+
+    auto exec = executor_factory.at(FLAGS_executor)();
+    auto engine = get_engine();
+    auto formats = split(FLAGS_formats, ',');
+
+    rapidjson::IStreamWrapper jcin(std::cin);
+    rapidjson::Document test_cases;
+    test_cases.ParseStream(jcin);
+    if (!test_cases.IsArray()) {
+        print_config_error_and_exit();
+    }
+
+    auto &allocator = test_cases.GetAllocator();
+
+    for (auto &test_case : test_cases.GetArray()) {
+        try {
+            // set up benchmark
+            validate_option_object(test_case);
+            if (!test_case.HasMember("spmv")) {
+                test_case.AddMember("spmv",
+                                    rapidjson::Value(rapidjson::kObjectType),
+                                    allocator);
+            }
+            auto &spmv_case = test_case["spmv"];
+            if (!FLAGS_overwrite &&
+                all_of(begin(formats), end(formats),
+                       [&spmv_case](const std::string &s) {
+                           return spmv_case.HasMember(s.c_str());
+                       })) {
+                continue;
+            }
+            std::clog << "Running test case: " << test_case << std::endl;
+            std::ifstream mtx_fd(test_case["filename"].GetString());
+            auto data = gko::read_raw<etype>(mtx_fd);
```

Comment:
I guess you're currently working on this. Maybe we can have new read_raw that takes a file name prefix, appends numeric suffixes to it to read each of the batch entries (assuming the file names are like that), and then returns a vector<matrix_data>?

Or, we can have a descriptor file that has the metadata for the batch, and we point read_raw to that file. That file can then be parsed to get the number of small matrices in the batch and the prefix for files containing each of the matrices.

Reply:
Yeah, I was planning on doing more or less the above. Now we populate the input file class and read the `matrix_data`. And we can either benchmark by duplicating either the input batch matrices as well or by using ssget and duplicating that.